integration-libs/epd-visualization/components/visual-viewer/visual-viewer.service.ts

```diff
@@ -0,0 +1,1498 @@
+import {
+  ChangeDetectorRef,
+  ElementRef,
+  EventEmitter,
+  Injectable,
+  OnDestroy,
+} from '@angular/core';
+// @ts-ignore
+import * as ui5 from '@sapui5/ts-types-esm';
+import {
+  SceneNodeToProductLookupService,
+  VisualizationLookupService,
+} from '@spartacus/epd-visualization/core';
+import {
+  ContentType,
+  EpdVisualizationConfig,
+  EpdVisualizationInnerConfig,
+  Ui5Config,
+  VisualizationApiConfig,
+  VisualizationInfo,
+} from '@spartacus/epd-visualization/root';
+import { BehaviorSubject, Observable, of, Subscription } from 'rxjs';
+import {
+  catchError,
+  filter,
+  first,
+  mergeMap,
+  shareReplay,
+} from 'rxjs/operators';
+import type Core from 'sap/ui/core/Core';
+import type { CSSColor } from 'sap/ui/core/library';
+import type UIArea from 'sap/ui/core/UIArea';
+import type AnimationPlayer from 'sap/ui/vk/AnimationPlayer';
+import type ContentConnector from 'sap/ui/vk/ContentConnector';
+import type ContentResource from 'sap/ui/vk/ContentResource';
+import type DrawerToolbar from 'sap/ui/vk/DrawerToolbar';
+import NodeHierarchy from 'sap/ui/vk/NodeHierarchy';
+import type Scene from 'sap/ui/vk/Scene';
+import type Viewport from 'sap/ui/vk/Viewport';
+import type ViewStateManager from 'sap/ui/vk/ViewStateManager';
+import { NavigationMode } from './models/navigation-mode';
+import { NodeContentType } from './models/node-content-type';
+import { LoadedSceneInfo, SceneLoadInfo } from './models/scene-load-info';
+import { SceneLoadState } from './models/scene-load-state';
+import { SelectionMode } from './models/selection-mode';
+import {
+  VisualizationLoadInfo,
+  VisualizationLoadStatus,
+  VisualizationLookupResult,
+} from './models/visualization-load-info';
+import { ZoomTo } from './models/zoom-to';
+
+type ViewManager = any;
+type NodeRef = any;
+
+interface VisualContentChangesFinishedEvent {
+  content: any;
+  failureReason: any;
+}
+
+interface VisualContentLoadFinishedEvent {}
+
+@Injectable({
+  providedIn: 'any',
+})
+export class VisualViewerService implements OnDestroy {
+  constructor(
+    protected epdVisualizationConfig: EpdVisualizationConfig,
+    protected _sceneNodeToProductLookupService: SceneNodeToProductLookupService,
+    protected visualizationLookupService: VisualizationLookupService,
+    protected elementRef: ElementRef,
+    protected changeDetectorRef: ChangeDetectorRef
+  ) {
+    const ui5BootStrapped$: Observable<void> =
+      this.bootstrapUi5('ui5bootstrap');
+    const ui5Initialized$: Observable<void> = ui5BootStrapped$.pipe(
+      mergeMap(this.initializeUi5.bind(this))
+    );
+    this.viewportAdded$ = ui5Initialized$.pipe(
+      mergeMap(this.addViewport.bind(this)),
+      shareReplay()
+    );
+    this.executeWhenSceneLoaded(this.setInitialPropertyValues.bind(this));
+  }
+
+  ngOnDestroy(): void {
+    this.selectedNodeIdsSubscription?.unsubscribe();
+  }
+
+  protected _ui5: ui5;
```

Comment:
Is this property really needed? I cannot find it being used anywhere.

Reply:
@sam-garland 
Maybe I'm missing something... but removing this property **looks very promising**, becasue it helped me avoid setting `skipLibCheck` in the customer's app OOTB! `skipLibCheck` is still need it in our monorepo app's tsconfig. But we could avoid it at least for customers by default, which I find a big win.
- When I removed `protected _ui5: ui5;`, compiled this lib, and published to local npm registry, and then installed in a fresh local Angular app (aka customer's app), and I disabled `skipLibCheck` in customer's app, it worked! I could run `yarn start` in customer's app successfully without errors in `visual-viewer.service.d.ts` or in `@sapui5/ts-types-esm` types! I also did a smoke QA test the feature looks and works correctly:
  <img width="1425" alt="image" src="https://user-images.githubusercontent.com/4001059/146232953-1a01e550-a61e-4908-a689-c77fbd0ea5e1.png">

- Even, when I removed the property `protected _ui5: ui5;`, I still need `skipLibCheck` in our monorepo tsconfig. (without `skipLibCheck`, UI5 types raise errors on `yarn start`). But I've run `yarn start` in our monorepo (with b2b,epd-visualisation enabled in env-cmd) and the feature seems to work without runtime errors:
  <img width="1431" alt="image" src="https://user-images.githubusercontent.com/4001059/146231369-4d892a06-faf4-4bc0-83e1-a31235acf4fd.png">


That said, would you have anything against removing the property `protected _ui5: ui5;`?

Reply:
The reason why this property exists is that without this the VSCode organise imports feature (that is performed on save when this setting is used is used will remove the `import * as ui5 from '@sapui5/ts-types-esm';` line resulting in a compilation failure:
https://github.com/SAP/spartacus/blob/63c3432f29e38d0b40a4825113e03e2042bec19c/.vscode/recommended-settings.json#L4

Reply:
I answered in the comment https://github.com/SAP/spartacus/pull/14420#discussion_r770096985. I'm sorry for confusing the thread.

Reply:
Understood, organizeImports can be annoying in this case. But IMHO avoiding this inconvenience for Spartacus developers is not enough reason to force all customers using the feature to degrade type-safety by setting `skipLibCheck` for them.

As an alternative to a fake import `import * as ui5 from '@sapui5/ts-types-esm'` (which is semantically incorrect, as '@sapui5/ts-types-esm' is not a module), did you consider declaring a [dependency on a package](https://www.typescriptlang.org/docs/handbook/triple-slash-directives.html#-reference-types-) in the top of file? It could be preceded with an explanation comment:
```
// '@sapui5/ts-types-esm' package contains types for sap modules, e.g. 'sap/ui/core/Core'
/// <reference types="@sapui5/ts-types-esm" />
```

--
I've just checked it works: the lib compiles successfully and I don't need `skipLibCheck` in customer's app.

Reply:
Thanks Krzysztof. I am also trying this now and figuring out the exact set of changes needed.

Reply:
The above solution with `/// <reference types="@sapui5/ts-types-esm" />`:
- works in our repo, but only with `skipLibCheck` enabled
- works in customer's app even without `skipLibCheck`

That said, if customers want to develop a customization of our service, they will likely want to use the original UI5 types in their custom source code. In that case, they will need to do the following: set `skipLibCheck` in their app AND add a TS directive `///<reference types>` in the top of their custom service file.

Reply:
OK, so I will:
- Remove the `import * as ui5 from '@sapui5/ts-types-esm` and `_ui5` field
- Remove the code from the epd-visualization schematic that was adding `skipLibCheck` to the target tsconfig file.
- Leave the `skipLibCheck` configuration as it is in the Spartacus repo

Are you happy with that?

Reply:
Correct. Yes, I'm happy 😉

Reply:
I made some analysis to explain the possible reason why `skipLibCheck` is required in our repo, but not in customer's repo. Let me share my observations:

- in our source file `visual-viewer.service.TS` we're depending directly on the d.ts types from `@sapui5/ts-types-esm`. Since they contain errors, we need to set `skipLibCheck` to make our library build to succeed.

- after building our library, the build artifact `dist/epd-visualization/components/visual-viewer/visual-viewer.service.d.ts` has no notion about the `@sapui5/ts-types-esm` package. See my gist https://gist.github.com/Platonn/b657572f64048c0024470dc3fa9a9095.

   Note: Just for the comparison, I created also a gist of `visual-viewer.service.d.ts` of the former implementation (with `_ui5: ui5` property and `import * as ui5`): https://gist.github.com/Platonn/531e7f27b466af2e44fe633a1d2d201b - the `import * as ui5` is leaked to the service's d.ts file

- eventually, when customers compiles their app, their source code depends directly only on our `visual-viewer.service.d.ts` file, which has notion about `@sapui5/ts-types-esm`. Therefore customers' build don't need `skipLibCheck`, becasue all d.ts included in their build are correct.

--
Note: I realized that if we change any service's property (with a type from `@sapui5/ts-types-esm`) from private to protected/public (by example `protected _viewport: Viewport`), customers will be forced to set `skipLibCheck` when _using_ our lib OOTB!

   Explanation: since all the service's properties are **private** , after the compilation of the lib, their types are stripped and `visual-viewer.service.d.ts` has no notion about `@sapui5/ts-types-esm`.
However if we make **any** of those properties **protected**, then the notion about `@sapui5/ts-types-esm` will leak into the d.ts and it will enforce using `skipLibCheck` OOTB in customer's codebase.

   To prove that, as an experiment, I made the `_viewport: Viewport` property `protected`. Then the customer's app I had to set `skipLibCheck` to avoid errors OOTB. Here I share the gist of `visual-viewer.service.d.ts` https://gist.github.com/Platonn/3bc1c63da489f9b4a5f918180d3d17b9 - `/// <reference types="@sapui5/ts-types-esm" />` leaked to the d.ts to help resolve the type `protected _viewport: Viewport`

Reply:
Currently all the properties are private, so it's not convenient to extend the service from customer's point of view. That's not an issue, if customers are not likely to extend it...

Having all those properties private is a nice situation for us (Spartacus developers), because the notion of `@sapui5/ts-types-esm` is stripped from our output `visual-viewer.service.d.ts` and we don't need to force all customers using EPD Visualisation to set `skipLibCheck`.

If in the future you discover that this service should be easier extendable and we should make all properties protected, then we might be forced to come back to the solution with setting `skipLibCheck` via schematics on behalf of customers 🤷‍♂️ 

All comes down to the needs of customers - whether they will likely need to customize this service or not. You and you team might know the answer to this question better.

Reply:
It is a bit of an accidental situation that all the references to UI5 types are private, but it is a lucky situation for now.
I think it makes sense to keep it that way and not have any UI5 types exposed via the public API (in most cases it should be possible to use types defined in @spartacus/epd-visualization that are mapped to UI5 types instead).

I think it will be a rare case for a customer to want/need to directly modify the code that is interacting with the viewer themselves. I think they are probably more likely to change theming, default values or modify the other components in the library.

Reply:
My changes relating to this have been merged.

Reply:
Thanks for your code updates!

PS. if by any chance in the future we need to get back to the solution with schematics adding the `skipLibCheck` in customer's app, I suggest to add it only to the customer's `tsconfig.app.json`, but not the global `tsconfig` of the customer's (mono)repo. They might have their own libraries having tsconfigs deriving from the root one, and perhaps not all of them will need `skipLibCheck`.

See the structure of a fresh Angular app: 
![image](https://user-images.githubusercontent.com/4001059/146584218-10bf262c-e4c2-4d02-9372-a4f259870888.png)